pennylane/tape/tape.py

```diff
@@ -149,37 +179,11 @@ def stop_at(obj):  # pylint: disable=unused-argument
     if tape.samples_computational_basis and len(tape.measurements) > 1:
         _validate_computational_basis_sampling(tape.measurements)
 
-    if tape._obs_sharing_wires:
-        with QueuingManager.stop_recording():  # stop recording operations to active context when computing qwc groupings
-            try:
-                rotations, diag_obs = qml.pauli.diagonalize_qwc_pauli_words(tape._obs_sharing_wires)
-                tape._obs_sharing_wires = diag_obs
-            except (TypeError, ValueError) as e:
-                if any(
-                    isinstance(m, (ProbabilityMP, SampleMP, CountsMP)) for m in tape.measurements
-                ):
-                    raise qml.QuantumFunctionError(
-                        "Only observables that are qubit-wise commuting "
-                        "Pauli words can be returned on the same wire.\n"
-                        "Try removing all probability, sample and counts measurements "
-                        "this will allow for splitting of execution and separate measurements "
-                        "for each non-commuting observable."
-                    ) from e
-
-                raise qml.QuantumFunctionError(
-                    _err_msg_for_some_meas_not_qwc(tape.measurements)
-                ) from e
-
-            tape._ops.extend(rotations)
-
-            for o, i in zip(diag_obs, tape._obs_sharing_wires_id):
-                new_m = tape.measurements[i].__class__(obs=o)
-                tape._measurements[i] = new_m
-
+    tape_ops, tape_measurements = get_ops_with_rotations_and_measurements(tape)
```

Comment:
I like the idea of splitting this logic into its own function :)

How about a call signature of:
```
diagonalizing_gates, new_measurements = rotations_and_diagonal_measurements(tape.measurements)
```
Then we use `tape._ops + diagonalizing_gates` instead of `tape_ops`?

I think that would make it more clear that this chunk of code is only occupied with measurements and diagonalizing gates, not the rest of the gates as whole.

Reply:
I was trying to do something like that but got caught up in the in-place measurement changes.. that is much better! I'll update accordingly.